server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java

```diff
@@ -0,0 +1,441 @@
+    {
+      Interval interval = getBucketInterval(segment);
+      buckets.computeIfPresent(
+          interval,
+          (i, builder) -> builder.removeSegment(segment).isEmpty() ? null : builder
+      );
+      return this;
+    }
+
+    public boolean isEmpty()
+    {
+      return buckets.isEmpty();
+    }
+
+    public SegmentsCostCache build()
+    {
+      return new SegmentsCostCache(
+          buckets
+              .entrySet()
+              .stream()
+              .map(entry -> entry.getValue().build())
+              .collect(Collectors.toCollection(ArrayList::new))
+      );
+    }
+
+    private static Interval getBucketInterval(DataSegment segment)
+    {
+      return BUCKET_GRANULARITY.bucket(segment.getInterval().getStart());
+    }
+  }
+
+  static class Bucket
+  {
+    private final Interval interval;
+    private final Interval calculationInterval;
+    private final ArrayList<DataSegment> sortedSegments;
+    private final double[] leftSum;
+    private final double[] rightSum;
+
+    Bucket(Interval interval, ArrayList<DataSegment> sortedSegments, double[] leftSum, double[] rightSum)
+    {
+      this.interval = Preconditions.checkNotNull(interval, "interval");
+      this.sortedSegments = Preconditions.checkNotNull(sortedSegments, "sortedSegments");
+      this.leftSum = Preconditions.checkNotNull(leftSum, "leftSum");
+      this.rightSum = Preconditions.checkNotNull(rightSum, "rightSum");
+      Preconditions.checkArgument(sortedSegments.size() == leftSum.length && sortedSegments.size() == rightSum.length);
+      Preconditions.checkArgument(SEGMENT_ORDERING.isOrdered(sortedSegments));
+      this.calculationInterval = new Interval(
+          interval.getStart().minus(LIFE_THRESHOLD),
+          interval.getEnd().plus(LIFE_THRESHOLD)
+      );
+    }
+
+    Interval getInterval()
+    {
+      return interval;
+    }
+
+    boolean inCalculationInterval(DataSegment dataSegment)
+    {
+      return calculationInterval.overlaps(dataSegment.getInterval());
+    }
+
+    double cost(DataSegment dataSegment)
+    {
+      // cost is calculated relatively to bucket start (which is considered as 0)
+      double t0 = convertStart(dataSegment, interval);
+      double t1 = convertEnd(dataSegment, interval);
+
+      // avoid calculation for segments outside of LIFE_THRESHOLD
+      if (!inCalculationInterval(dataSegment)) {
+        throw new ISE("Segment is not within calculation interval");
+      }
+
+      int index = Collections.binarySearch(sortedSegments, dataSegment, SEGMENT_INTERVAL_COMPARATOR);
+      index = (index >= 0) ? index : -index - 1;
+      return addLeftCost(dataSegment, t0, t1, index) + rightCost(dataSegment, t0, t1, index);
+    }
+
+    private double addLeftCost(DataSegment dataSegment, double t0, double t1, int index)
+    {
+      double leftCost = 0.0;
+      // add to cost all left-overlapping segments
+      int leftIndex = index - 1;
+      while (leftIndex >= 0
```

Comment:
This loop looks pretty odd, can it simply be

```java
for (int leftIndex = index - 1; leftIndex >=0; --leftIndex) {
final Segment segment = sortedSegments.get(leftIndex);
if (!segment.getInterval().overlaps(dataSegment.getInterval()))
{
  break;
}
final double start = convertStart(sortedSegments.get(leftIndex), interval);
final double end = convertEnd(sortedSegments.get(leftIndex), interval);
leftCost += CostBalancerStrategy.intervalCost(end - start, t0 - start, t1 - start);
}
```
or similar?

Reply:
It uses `leftCost` variable outside of the loop